### ISSUE TITLE
[Request]: Pull/Scroll down to refresh

### ISSUE THREAD

**Issue by AnuraagReddy123** [GeopJr/Tuba#1004]
### Describe the request

Apps like Reddit, Mail in iOS have this feature, that refreshes the app once you pull down on the screen. 
<img src="https://github.com/GeopJr/Tuba/assets/66117695/6d1c275a-3ad3-4011-9572-62c987829a1a" width="500">

A feature like this would be nice to have in Tuba but I am not sure how feasible it would be to implement. I think the app detects when the user has reached the top of the page, as an animation plays out when you reach the top. So maybe the app can be refreshed when that animation is completed?

https://github.com/GeopJr/Tuba/assets/66117695/01e0500a-69c5-4ef7-9842-21ab934074eb

### Implementation Details

- [X] This follows the [GNOME HIG](https://developer.gnome.org/hig/).

**Comment by GeopJr** [comment]
Thanks for raising this issue!

Tuba can do this already!

[Screencast from 2024-06-14 10-49-05.webm](https://github.com/GeopJr/Tuba/assets/18014039/bce17e8c-5b3a-42e8-b54f-edbeb1f8a64d)

But by clicking/tapping and dragging down.

Lengthy explanation on why it works like this:

TLDR: We've tested out a lot of different ways to do this and this was the most reliable. That's not to say that it cannot change in the future, but for now this is the best of all worlds.

Pull to refresh was first suggested on #155. This is a new pattern / not found in any other GNOME app so it was up to me to figure out all the details.

In my comment https://github.com/GeopJr/Tuba/issues/155#issuecomment-1560440341, you'll see my first implementation. It worked similar to what you suggested, if the user scrolled 'harder' at the top, it would show a spinner, and if they scrolled to the top again, it would refresh.

![](https://private-user-images.githubusercontent.com/18014039/240482373-fb5ffaba-5937-460f-bf32-c633a47298df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNTE4NTQsIm5iZiI6MTcxODM1MTU1NCwicGF0aCI6Ii8xODAxNDAzOS8yNDA0ODIzNzMtZmI1ZmZhYmEtNTkzNy00NjBmLWJmMzItYzYzM2E0NzI5OGRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE0VDA3NTIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzNDgyMGEwNWRjNWYxODVmNmRmNjIxZjZjZjYzOGVkMTM1MjgwNTdkOTllOTczNzYzZDNkNjMyMDdlNTY0YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LfUtoKd198eOZJh7YmEZgiyCKQXCSbwlu1bFPXuYJis)

https://private-user-images.githubusercontent.com/18014039/240482709-16dddacf-f524-4220-8655-4a10754b8c86.webm?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNTE4NTQsIm5iZiI6MTcxODM1MTU1NCwicGF0aCI6Ii8xODAxNDAzOS8yNDA0ODI3MDktMTZkZGRhY2YtZjUyNC00MjIwLTg2NTUtNGExMDc1NGI4Yzg2LndlYm0_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxNFQwNzUyMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xM2YwOWE0ZjBjOGNiMDAyNGFkYjAwOTc3OGYxMjFmMDIxYzRlOGFiMTZkMzU1YmFhOTM2Y2VkNGI1MWFlMmFkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wt01cc7b5QtpiGUEzB2Efcj5bqCW577Gn6zSKFcBdzk


That however did not feel as natural. It was hard to discover and even hard to trigger, plus it required a scroll wheel.

So instead on #283, I changed it to the 'drag' behavior.

Design wise, it was similar to apple's

![](https://private-user-images.githubusercontent.com/18014039/254722577-d78f8c23-892a-4cf3-8602-f3552eb00678.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNTE4NDMsIm5iZiI6MTcxODM1MTU0MywicGF0aCI6Ii8xODAxNDAzOS8yNTQ3MjI1NzctZDc4ZjhjMjMtODkyYS00Y2YzLTg2MDItZjM1NTJlYjAwNjc4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE0VDA3NTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwODY3OWUwNDU5NjcyYjZkNDQ4OWEzYTJjNWI4ZWE2YzA4YzZmYTc1ZmM0YmRiNDMxNWJkMDc4NzU1MzA0ODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ryfk6m-f9AWRO3IJkHreEn0VWaFHUw6PJ-LEwSjpJdc)

But later due to some performance experiments, it changed to the 'android'-style one (the one you see at the first recording in this comment)
